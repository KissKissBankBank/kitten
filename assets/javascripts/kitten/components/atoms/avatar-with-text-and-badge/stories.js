import React from 'react'
import { AvatarWithTextAndBadge } from './index'
import { Text, COLORS } from '../../..'

export default {
  title: 'Atoms/AvatarWithTextAndBadge',
  component: AvatarWithTextAndBadge,
  decorators: [
    story => (
      <div className="story-Container story-Grid">
        <div>{story()}</div>
      </div>
    ),
  ],
  args: {
    hoverable: true,
    imageAlt: 'Image alt',
    imageSize: 'regular',
    imageIsDisplayed: true,
    imageText: 'JC',
    imageBackgroundColor: '#002e7d',
    imageTextColor: '#fff',
    badgeValue: 0,
    badgeBackgroundColor: '#002e7d',
    badgeA11yText: 'notifications',
    textContent: 'Jean Charles Édouard',
    textHasSecondLine: false,
    textWithEllipsisOverflow: false,
  },
  argTypes: {
    hoverable: {
      name: 'Hoverable',
      defaultValue: false,
      control: { type: 'boolean' },
    },
    imageAlt: {
      name: 'Image Alt (story prop)',
      defaultValue: null,
      description: 'Prop `alt` on subcomponent `AvatarWithTextAndBadge.Image`',
      control: { type: 'text' },
    },
    imageSize: {
      name: 'Image Size (story prop)',
      defaultValue: 'regular',
      description: 'Prop `size` on subcomponent `AvatarWithTextAndBadge.Image`',
      options: ['regular', 'big'],
      control: { type: 'inline-radio' },
    },
    imageIsDisplayed: {
      name: 'Image Is Displayed (story prop)',
      defaultValue: null,
      description:
        'Controls the prop `src` on subcomponent `AvatarWithTextAndBadge.Image`',
      control: { type: 'boolean' },
    },
    imageText: {
      name: 'Image Text (story prop)',
      defaultValue: '',
      description: 'Prop `text` on subcomponent `AvatarWithTextAndBadge.Image`',
      control: { type: 'text' },
    },
    imageBackgroundColor: {
      name: 'Image BackgroundColor (story prop)',
      defaultValue: COLORS.primary3,
      description:
        'Prop `backgroundColor` on subcomponent `AvatarWithTextAndBadge.Image`',
      control: { type: 'color' },
    },
    imageTextColor: {
      name: 'Image TextColor (story prop)',
      defaultValue: COLORS.background1,
      description:
        'Prop `textColor` on subcomponent `AvatarWithTextAndBadge.Image`',
      control: { type: 'color' },
    },
    badgeValue: {
      name: 'Badge number (story prop)',
      defaultValue: null,
      description:
        'Children of the subcomponent `AvatarWithTextAndBadge.Badge`',
      control: { type: 'number' },
    },
    badgeBackgroundColor: {
      name: 'Badge backgroundColor (story prop)',
      defaultValue: COLORS.primary1,
      description:
        'Prop `backgroundColor` of subcomponent `AvatarWithTextAndBadge.Badge`',
      control: { type: 'color' },
    },
    badgeA11yText: {
      name: 'Badge a11yText (story prop)',
      defaultValue: 'Notification(s)',
      description:
        'Prop `a11yText` of subcomponent `AvatarWithTextAndBadge.Badge`',
      control: { type: 'text' },
    },
    textContent: {
      name: 'Text content (story prop)',
      defaultValue: '',
      description: 'Children of the subcomponent `AvatarWithTextAndBadge.Text`',
      control: { type: 'text' },
    },
    textHasSecondLine: {
      name: 'Text display second line (story prop)',
      description:
        'Adds complex children to the subcomponent `AvatarWithTextAndBadge.Text`',
      control: { type: 'boolean' },
    },
    textWithEllipsisOverflow: {
      name: 'Text has overflow ellipsis (story prop)',
      defaultValue: false,
      description:
        'Prop `withEllipsisOverflow` of subcomponent `AvatarWithTextAndBadge.Text`',
      control: { type: 'boolean' },
    },
  },
}

export const Default = ({
  hoverable,
  imageAlt,
  imageSize,
  imageIsDisplayed,
  imageText,
  imageBackgroundColor,
  imageTextColor,
  badgeValue,
  badgeBackgroundColor,
  badgeA11yText,
  textContent,
  textHasSecondLine,
  textWithEllipsisOverflow,
}) => (
  <AvatarWithTextAndBadge hoverable={hoverable}>
    <AvatarWithTextAndBadge.Image
      alt={imageAlt}
      size={imageSize}
      src={imageIsDisplayed && '/kitten.jpg'}
      text={imageText}
      backgroundColor={imageBackgroundColor}
      textColor={imageTextColor}
    >
      {badgeValue > 0 && (
        <AvatarWithTextAndBadge.Badge
          backgroundColor={badgeBackgroundColor}
          a11yText={badgeA11yText}
          children={badgeValue}
        />
      )}
    </AvatarWithTextAndBadge.Image>

    <AvatarWithTextAndBadge.Text
      className="k-u-hidden@xxs-down"
      withEllipsisOverflow={textWithEllipsisOverflow}
    >
      {textContent}
      {textHasSecondLine && (
        <>
          <br />
          <Text weight="light" size="tiny">
            <span className="k-u-hidden@xs-down">Solde&nbsp;:</span>{' '}
            24&nbsp;093,39&nbsp;€
          </Text>
        </>
      )}
    </AvatarWithTextAndBadge.Text>
  </AvatarWithTextAndBadge>
)
