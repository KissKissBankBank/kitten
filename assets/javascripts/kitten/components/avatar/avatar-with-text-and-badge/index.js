import React from 'react'
import styled from 'styled-components'
import classNames from 'classnames'
import PropTypes from 'prop-types'
import { Text as KittenText } from '../../../components/typography/text'
import { Badge as KittenBadge } from '../../../components/atoms/badge'
import COLORS from '../../../constants/colors-config'
import { pxToRem } from '../../../helpers/utils/typography'
import { getReactElementsByType } from '../../../helpers/react/react-elements'

const WithoutAvatar = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;

  color: ${({ color }) => color};
  background-color: ${({ backgroundColor }) => backgroundColor};
  width: ${({ width }) => width};
  height: ${({ height }) => height};
`

const Image = ({
  src,
  height,
  width,
  textColor,
  backgroundColor,
  alt,
  children,
  text,
}) => {
  const badgeElement = getReactElementsByType({
    children,
    type: AvatarWithTextAndBadge.Badge,
  })

  const badge = badgeElement && badgeElement[0]

  return (
    <span className="k-ButtonWithBadge">
      <span className="k-ButtonImage">
        {src ? (
          <img
            className="k-ButtonImage__img"
            src={src}
            width={width}
            height={height}
            alt={alt}
          />
        ) : (
          <WithoutAvatar
            color={textColor}
            backgroundColor={backgroundColor}
            width={width}
            height={height}
          >
            <KittenText size="micro" weight="regular">
              {text}
            </KittenText>
          </WithoutAvatar>
        )}
      </span>
      {badge}
    </span>
  )
}

Image.propTypes = {
  src: PropTypes.string,
  height: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  width: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  textColor: PropTypes.string,
  backgroundColor: PropTypes.string,
  alt: PropTypes.string,
}

Image.defaultProps = {
  src: null,
  height: pxToRem(40),
  width: pxToRem(40),
  textColor: COLORS.background1,
  backgroundColor: COLORS.primary3,
  alt: null,
}

const Badge = ({ backgroundColor, children, a11yText }) => (
  <KittenBadge
    className="k-ButtonWithBadge__badge"
    color={backgroundColor}
    a11yText={a11yText}
  >
    {children}
  </KittenBadge>
)

Badge.propTypes = {
  backgroundColor: PropTypes.string,
  a11yText: PropTypes.string,
}

Badge.defaultProps = {
  backgroundColor: COLORS.primary1,
  a11yText: 'Notification(s)',
}

const Text = ({ textClassName, withEllipsisOverflow, children }) => {
  const textClassNames = classNames(
    textClassName,
    'k-ButtonImageWithText__text',
    'k-ButtonImageWithText__text--withoutPaddingRight',
    'always-visible',
    { 'text--withEllipsis': withEllipsisOverflow },
  )

  return <span className={textClassNames}>{children}</span>
}

Text.propTypes = {
  withEllipsisOverflow: PropTypes.bool,
}

Text.defaultProps = {
  withEllipsisOverflow: false,
}

export const AvatarWithTextAndBadge = ({ children, className, ...props }) => {
  const componentClassNames = classNames(
    className,
    'k-ButtonImageWithText',
    'k-u-avoid-click',
  )

  return (
    <div className={componentClassNames} {...props}>
      {children}
    </div>
  )
}

AvatarWithTextAndBadge.Image = Image
AvatarWithTextAndBadge.Badge = Badge
AvatarWithTextAndBadge.Text = Text
