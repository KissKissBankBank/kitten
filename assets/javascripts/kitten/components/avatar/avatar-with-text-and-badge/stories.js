import React from 'react'
import { AvatarWithTextAndBadge } from './index'
import { text, color } from '@storybook/addon-knobs'
import { Text } from '../../../components/typography/text'

export default {
  title: 'Avatar/AvatarWithTextAndBadge',
  component: AvatarWithTextAndBadge,
  decorators: [
    Story => (
      <div className="story-Container">
        <Story />
      </div>
    ),
  ],
}

export const WithAvatar = args => (
  <AvatarWithTextAndBadge {...args}>
    <AvatarWithTextAndBadge.Image
      src={text('Image src', 'https://via.placeholder.com/40x40.png')}
      alt={text('Image Alt', 'Avatar')}
    >
      <AvatarWithTextAndBadge.Badge
        a11yText={text('Accessibility Badge text', '2 notifications')}
        children={text('Badge value', '2')}
      />
    </AvatarWithTextAndBadge.Image>

    <AvatarWithTextAndBadge.Text
      textClassName="k-u-hidden@xxs-down"
      withEllipsisOverflow={true}
    >
      <Text
        lineHeight="normal"
        weight="regular"
        children={text('Text', 'Jean Charles Édouard')}
      />
    </AvatarWithTextAndBadge.Text>
  </AvatarWithTextAndBadge>
)

export const WithText = args => (
  <AvatarWithTextAndBadge {...args}>
    <AvatarWithTextAndBadge.Image
      text={text('', 'JC')}
      backgroundColor={color('Avatar background color', '#002e7d')}
      textColor={color('Text color', '#fff')}
    >
      <AvatarWithTextAndBadge.Badge
        a11yText={text('Accessibility Badge text', '2 notifications')}
        children={text('Badge value', '2')}
        backgroundColor={color('Badge color', '#19b4fa')}
      />
    </AvatarWithTextAndBadge.Image>

    <AvatarWithTextAndBadge.Text
      textClassName="k-u-hidden@xxs-down"
      withEllipsisOverflow={true}
    >
      <Text lineHeight="normal" weight="regular">
        Jean Charles Édouard
      </Text>
      <br />
      <Text weight="light">
        <Text className="k-u-weight-light k-u-hidden@xs-down">
          Solde&nbsp;:
        </Text>{' '}
        24&nbsp;093,39&nbsp;€
      </Text>
    </AvatarWithTextAndBadge.Text>
  </AvatarWithTextAndBadge>
)
