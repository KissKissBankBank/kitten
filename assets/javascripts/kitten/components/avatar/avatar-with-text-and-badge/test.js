import React from 'react'
import renderer from 'react-test-renderer'
import { AvatarWithTextAndBadge } from './index'
import { Text } from '../../../components/typography/text'

describe('<AvatarWithTextAndBadge />', () => {
  describe('with avatar', () => {
    it('should match with snapshot', () => {
      const tree = renderer
        .create(
          <AvatarWithTextAndBadge
            imageProps={{
              src: 'https://via.placeholder.com/40x40.png',
              alt: 'Default image',
            }}
            notifications="5"
            text="Jean-Luc"
          />,
        )
        .toJSON()

      expect(tree).toMatchSnapshot()
    })
  })

  describe('without avatar', () => {
    it('should match with snapshot', () => {
      const tree = renderer
        .create(
          <AvatarWithTextAndBadge
            imageProps={{
              text: (
                <Text size="micro" weight="bold">
                  JC
                </Text>
              ),
              backgroundColor: '#002e7d',
            }}
            notifications="2"
            text={
              <Text lineHeight="normal" weight="bold">
                Jean Charles
              </Text>
            }
            subText={<Text weight="light">Solde : 1&nbsp;000,97&nbsp;â‚¬</Text>}
          />,
        )
        .toJSON()

      expect(tree).toMatchSnapshot()
    })
  })
})
