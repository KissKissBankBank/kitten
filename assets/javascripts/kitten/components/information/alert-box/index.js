import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { StatusIconNext } from '../../graphics/icons-next/status-icon-next'
import { IconBadge } from '../../information/icon-badge'
import TYPOGRAPHY from '../../../constants/typography-config'
import { pxToRem } from '../../../helpers/utils/typography'
import classNames from 'classnames'

const AlertBoxWrapper = styled.div`
  ${TYPOGRAPHY.fontStyles.light};
  border-radius: var(--border-radius-s, ${pxToRem(6)});
  overflow: hidden;
  background-color: var(--color-primary-100);
  color: var(--color-grey-900);
  gap: var(--alertBox-gap);
  padding: ${pxToRem(18)} var(--alertBox-gap);

  &.k-AlertBox--content {
    display: inline-flex;
  }

  &.k-AlertBox--fluid {
    display: flex;
  }

  &.k-AlertBox--icon-start {
    align-items: flex-start;
  }

  &.k-AlertBox--icon-center {
    align-items: center;
  }

  .k-AlertBox__icon {
    flex: 0 0 auto;
    border-radius: var(--border-radius-rounded, ${pxToRem(9999)});
  }

  .k-AlertBox__text {
    flex: 1 1 auto;
    font-size: ${pxToRem(14)};
    line-height: ${pxToRem(18)};
  }

  &.k-AlertBox--medium {
    gap: ${pxToRem(15)};
    padding: ${pxToRem(10)};

    .k-AlertBox__icon {
      width: ${pxToRem(20)};
      height: ${pxToRem(20)};
      flex-basis: ${pxToRem(20)};
    }
  }

  &.k-AlertBox--large {
    gap: ${pxToRem(20)};
    padding: ${pxToRem(20)};

    .k-AlertBox__icon {
      width: ${pxToRem(30)};
      height: ${pxToRem(30)};
      flex-basis: ${pxToRem(30)};
    }
  }

  &.k-AlertBox--small {
    gap: ${pxToRem(5)};
    padding: ${pxToRem(5)};

    .k-AlertBox__text {
      font-size: ${pxToRem(12)};
      line-height: ${pxToRem(15)};
    }
    .k-AlertBox__icon {
      width: ${pxToRem(16)};
      height: ${pxToRem(16)};
      flex-basis: ${pxToRem(16)};
    }
  }

  .k-u-link,
  a {
    ${TYPOGRAPHY.fontStyles.bold};
    text-decoration: none;
  }

  &.k-AlertBox--info {
    background-color: var(--color-primary-100);

    .k-u-link,
    a {
      color: var(--color-primary-500);
    }
  }

  &.k-AlertBox--success {
    background-color: var(--color-success-100);

    .k-u-link,
    a {
      color: var(--color-success-500);
    }
  }

  &.k-AlertBox--danger {
    background-color: var(--color-danger-100);

    .k-u-link,
    a {
      color: var(--color-danger-500);
    }
  }

  &.k-AlertBox--warning {
    background-color: var(--color-warning-100);

    .k-u-link,
    a {
      color: var(--color-warning-500);
    }
  }

  &.k-AlertBox--pending {
    background-color: var(--color-grey-300);

    .k-u-link,
    a {
      color: var(--color-primary-500);
    }
  }

  &.k-AlertBox--disabled {
    background-color: var(--color-grey-200);

    .k-u-link,
    a {
      color: var(--color-grey-600);
    }
  }
`

export const AlertBox = ({
  className,
  children,
  icon,
  status,
  displayIcon,
  iconPosition,
  size,
  fit,
  iconHasBorder,
  ...others
}) => {
  const internalIcon = (() => {
    if (icon) return icon

    return <StatusIconNext status={status} />
  })()

  const role = (() => {
    switch (status) {
      case 'danger':
        return 'alert'
      case 'warning':
      case 'success':
        return 'status'
      default:
        return null
    }
  })()

  const iconSize = (() => {
    switch (size) {
      case 'large':
        return 'medium'
      case 'medium':
        return 'small'
      case 'small':
        return 'micro'
    }
  })()

  return (
    <AlertBoxWrapper
      role={role}
      className={classNames(
        'k-AlertBox',
        className,
        `k-AlertBox--${status}`,
        `k-AlertBox--${size}`,
        `k-AlertBox--${fit}`,
        `k-AlertBox--icon-${iconPosition}`,
        {
          'k-AlertBox--hasIcon': displayIcon,
        },
      )}
      {...others}
    >
      {displayIcon && (
        <IconBadge
          className="k-AlertBox__icon"
          children={internalIcon}
          status={status}
          size={iconSize}
          hasBorder={iconHasBorder}
        />
      )}

      <div className="k-AlertBox__text">{children}</div>
    </AlertBoxWrapper>
  )
}

AlertBox.propTypes = {
  status: PropTypes.oneOf([
    'info',
    'success',
    'danger',
    'warning',
    'disabled',
    'pending',
  ]),
  displayIcon: PropTypes.bool,
  iconPosition: PropTypes.oneOf(['start', 'center']),
  icon: PropTypes.node,
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  fit: PropTypes.oneOf(['content', 'fluid']),
  iconHasBorder: PropTypes.bool,
}

AlertBox.defaultProps = {
  status: 'info',
  displayIcon: true,
  icon: null,
  size: 'medium',
  fit: 'fluid',
  iconPosition: 'start',
  iconHasBorder: true,
}
