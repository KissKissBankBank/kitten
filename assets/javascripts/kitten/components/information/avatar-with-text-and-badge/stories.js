import React from 'react'
import { AvatarWithTextAndBadge } from './index'
import { Text } from 'kitten'
import { DocsPage } from 'storybook/docs-page'

export default {
  title: 'Information/AvatarWithTextAndBadge',
  component: AvatarWithTextAndBadge,
  parameters: {
    docs: {
      page: () => (
        <DocsPage filepath={__filename} importString="AvatarWithTextAndBadge" />
      ),
    },
  },
  decorators: [
    story => (
      <div className="story-Container story-Grid">
        <div>{story()}</div>
      </div>
    ),
  ],
  args: {
    hoverable: true,
    imageAlt: 'Image alt',
    imageSize: 'medium',
    imageIsDisplayed: true,
    imageText: 'JC',
    imageBackgroundColor: '#002e7d',
    imageTextColor: '#fff',
    badgeValue: 0,
    badgeBackgroundColor: '#002e7d',
    badgeA11yText: 'notifications',
    textContent: 'Jean Charles Édouard',
    textHasSecondLine: false,
    textWithEllipsisOverflow: false,
  },
  argTypes: {
    hoverable: {
      name: 'Hoverable',
      control: 'boolean',
    },
    imageAlt: {
      name: 'Image Alt (story prop)',
      description: 'Prop `alt` on subcomponent `AvatarWithTextAndBadge.Image`',
      control: 'text',
    },
    imageSize: {
      name: 'Image Size (story prop)',
      description: 'Prop `size` on subcomponent `AvatarWithTextAndBadge.Image`',
      options: ['medium', 'large'],
      control: 'inline-radio',
    },
    imageIsDisplayed: {
      name: 'Image Is Displayed (story prop)',
      description:
        'Controls the prop `src` on subcomponent `AvatarWithTextAndBadge.Image`',
      control: 'boolean',
    },
    imageText: {
      name: 'Image Text (story prop)',
      description: 'Prop `text` on subcomponent `AvatarWithTextAndBadge.Image`',
      control: 'text',
    },
    imageBackgroundColor: {
      name: 'Image BackgroundColor (story prop)',
      description:
        'Prop `backgroundColor` on subcomponent `AvatarWithTextAndBadge.Image`',
      control: 'color',
    },
    imageTextColor: {
      name: 'Image TextColor (story prop)',
      description:
        'Prop `textColor` on subcomponent `AvatarWithTextAndBadge.Image`',
      control: 'color',
    },
    badgeValue: {
      name: 'Badge number (story prop)',
      description:
        'Children of the subcomponent `AvatarWithTextAndBadge.Badge`',
      control: 'number',
    },
    badgeBackgroundColor: {
      name: 'Badge backgroundColor (story prop)',
      description:
        'Prop `backgroundColor` of subcomponent `AvatarWithTextAndBadge.Badge`',
      control: 'color',
    },
    badgeA11yText: {
      name: 'Badge a11yText (story prop)',
      description:
        'Prop `a11yText` of subcomponent `AvatarWithTextAndBadge.Badge`',
      control: 'text',
    },
    textContent: {
      name: 'Text content (story prop)',
      description: 'Children of the subcomponent `AvatarWithTextAndBadge.Text`',
      control: 'text',
    },
    textHasSecondLine: {
      name: 'Text display second line (story prop)',
      description:
        'Adds complex children to the subcomponent `AvatarWithTextAndBadge.Text`',
      control: 'boolean',
    },
    textWithEllipsisOverflow: {
      name: 'Text has overflow ellipsis (story prop)',
      description:
        'Prop `withEllipsisOverflow` of subcomponent `AvatarWithTextAndBadge.Text`',
      control: 'boolean',
    },
  },
}

export const Default = ({
  hoverable,
  imageAlt,
  imageSize,
  imageIsDisplayed,
  imageText,
  imageBackgroundColor,
  imageTextColor,
  badgeValue,
  badgeBackgroundColor,
  badgeA11yText,
  textContent,
  textHasSecondLine,
  textWithEllipsisOverflow,
}) => (
  <AvatarWithTextAndBadge hoverable={hoverable}>
    <AvatarWithTextAndBadge.Image
      alt={imageAlt}
      size={imageSize}
      src={imageIsDisplayed && `/kitten-${Math.floor(Math.random() * 10)}.jpg`}
      text={imageText}
      backgroundColor={imageBackgroundColor}
      textColor={imageTextColor}
    >
      {badgeValue > 0 && (
        <AvatarWithTextAndBadge.Badge
          backgroundColor={badgeBackgroundColor}
          a11yText={badgeA11yText}
          children={badgeValue}
        />
      )}
    </AvatarWithTextAndBadge.Image>

    <AvatarWithTextAndBadge.Text
      className="k-u-hidden@xxs-down"
      withEllipsisOverflow={textWithEllipsisOverflow}
    >
      {textContent}
      {textHasSecondLine && (
        <>
          <br />
          <Text weight="light" size="small">
            <span className="k-u-hidden@xs-down">Solde&nbsp;:</span>{' '}
            24&nbsp;093,39&nbsp;€
          </Text>
        </>
      )}
    </AvatarWithTextAndBadge.Text>
  </AvatarWithTextAndBadge>
)
