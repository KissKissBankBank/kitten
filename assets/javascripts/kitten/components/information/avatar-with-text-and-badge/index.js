import React from 'react'
import styled from 'styled-components'
import classNames from 'classnames'
import PropTypes from 'prop-types'

import { Badge as BadgeComponent } from '../../information/badge'
import COLORS from '../../../constants/colors-config'
import { pxToRem, stepToRem } from '../../../helpers/utils/typography'
import TYPOGRAPHY from '../../../constants/typography-config'
import { getReactElementsByType } from '../../../helpers/react/get-react-elements'

const StyledWrapper = styled.div`
  display: flex;
  align-items: center;
  padding: 0;
  border: none;
  background: none;
  text-decoration: none;
  gap: ${pxToRem(10)};

  .k-Avatar {
    position: relative;
    padding: 0;
  }

  .k-Avatar__imageWrapper {
    width: ${pxToRem(40)};
    height: ${pxToRem(40)};
    border-radius: var(--border-radius-rounded);
    box-sizing: border-box;
    display: flex;
    overflow: hidden;
    border: var(--border-width) solid var(--color-grey-300);
    padding: 0;

    color: var(--k-Avatar-color);
    background-color: var(--k-Avatar-background-color);
  }

  .k-Avatar__image {
    display: block;
    margin: 0;
    padding: 0;
    border: 0;
    width: 100%;
    height: 100%;
  }

  .k-Avatar__noImage {
    display: flex;
    width: 100%;
    height: 100%;
    align-items: center;
    justify-content: center;

    ${TYPOGRAPHY.fontStyles['500']}
    font-size: ${stepToRem(-2)};
    line-height: 1;
  }

  .k-Avatar__badge {
    position: absolute;
    top: ${pxToRem(-4)};
    right: ${pxToRem(-8)};
    border: var(--border-width) solid white;
  }

  .k-Avatar__text {
    padding: pxToRem(10);
    padding-right: 0;
    text-align: left;
    color: ${COLORS.font1};
    text-decoration: none;

    ${TYPOGRAPHY.fontStyles['500']}
    font-size: ${stepToRem(-1)};
    line-height: 1.2;
  }

  .k-Avatar__text--hasEllipsis {
    &,
    & * {
      max-width: 100%;
      text-overflow: ellipsis;
      white-space: nowrap;
      overflow: hidden;
    }
  }

  .k-Avatar--small {
    .k-Avatar__imageWrapper {
      width: ${pxToRem(30)};
      height: ${pxToRem(30)};
    }

    .k-Avatar__noImage {
      font-size: ${stepToRem(-2)};
    }

    & + .k-Avatar__text {
      font-size: ${stepToRem(-2)};
    }
  }

  .k-Avatar--large {
    margin-right: ${pxToRem(15)};

    .k-Avatar__imageWrapper {
      width: ${pxToRem(60)};
      height: ${pxToRem(60)};
    }
    .k-Avatar__noImage {
      font-size: ${stepToRem(-1)};
    }

    & + .k-Avatar__text {
      font-size: ${stepToRem(0)};
    }
  }

  .k-Avatar--customSize {
    width: var(--Avatar-width);
    height: var(--Avatar-height);
  }

  a&,
  button&,
  &.k-Avatar__wrapper--isHoverable {
    cursor: pointer;

    .k-Avatar {
      opacity: 1;
      transition: opacity 0.2s;
    }

    &:hover .k-Avatar,
    &:focus .k-Avatar {
      opacity: 0.8;
      border: var(--border-hover);
    }
  }
`

const Image = ({
  alt,
  backgroundColor,
  children,
  className,
  height,
  size,
  src,
  style,
  text,
  textColor,
  width,
  ...props
}) => {
  const badgeElement = getReactElementsByType({
    children,
    type: AvatarWithTextAndBadge.Badge,
  })

  const badge = badgeElement && badgeElement[0]

  return (
    <div
      className={classNames('k-Avatar', className, `k-Avatar--${size}`, {
        'k-Avatar--customSize': width && height,
      })}
      style={{
        ...style,
        '--k-Avatar-color': textColor,
        '--k-Avatar-background-color': backgroundColor,
        '--Avatar-width': width,
        '--Avatar-height': height,
      }}
    >
      <span className="k-Avatar__imageWrapper">
        {src ? (
          <img className="k-Avatar__image" src={src} alt={alt} {...props} />
        ) : (
          <span className="k-Avatar__noImage">{text}</span>
        )}
      </span>
      {badge}
    </div>
  )
}

Image.propTypes = {
  src: PropTypes.string,
  height: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  width: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  textColor: PropTypes.string,
  backgroundColor: PropTypes.string,
  alt: PropTypes.string,
  size: PropTypes.oneOf(['small', 'medium', 'large']),
}

Image.defaultProps = {
  src: null,
  textColor: COLORS.background1,
  backgroundColor: COLORS.primary3,
  alt: '',
  size: 'medium',
}

const Badge = ({ backgroundColor, className, a11yText, ...props }) => (
  <BadgeComponent
    className={classNames('k-Avatar__badge', className)}
    color={backgroundColor}
    a11yText={a11yText}
    {...props}
  />
)

Badge.propTypes = {
  backgroundColor: PropTypes.string,
  a11yText: PropTypes.string,
}

Badge.defaultProps = {
  backgroundColor: COLORS.primary1,
  a11yText: 'Notification(s)',
}

const Text = ({ className, withEllipsisOverflow, ...props }) => {
  return (
    <span
      className={classNames('k-Avatar__text', className, {
        'k-Avatar__text--hasEllipsis': withEllipsisOverflow,
      })}
      {...props}
    />
  )
}

Text.propTypes = {
  withEllipsisOverflow: PropTypes.bool,
}

Text.defaultProps = {
  withEllipsisOverflow: false,
}

export const AvatarWithTextAndBadge = ({ className, hoverable, ...props }) => {
  return (
    <StyledWrapper
      className={classNames('k-Avatar__wrapper', className, {
        'k-Avatar__wrapper--isHoverable': hoverable,
      })}
      {...props}
    />
  )
}

AvatarWithTextAndBadge.propTypes = {
  hoverable: PropTypes.bool,
}

AvatarWithTextAndBadge.defaultProps = {
  hoverable: false,
}

AvatarWithTextAndBadge.Image = Image
AvatarWithTextAndBadge.Badge = Badge
AvatarWithTextAndBadge.Text = Text
