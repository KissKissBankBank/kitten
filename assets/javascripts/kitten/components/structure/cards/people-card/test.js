import React from 'react'
import renderer from 'react-test-renderer'
import { PeopleCard } from './index'
import { AvatarWithTextAndBadge } from '../../../information/avatar-with-text-and-badge'

describe('<PeopleCard />', () => {
  let component

  describe('default', () => {
    component = renderer
      .create(
        <PeopleCard>
          <PeopleCard.Image />
          <PeopleCard.Cell>
            Fusce dapibus, tellus ac cursus commodo, tortor mauris condimentum
            nibh, ut fermentum massa justo sit amet risus.
          </PeopleCard.Cell>
          <PeopleCard.LastCell>•••</PeopleCard.LastCell>
        </PeopleCard>,
      )
      .toJSON()

    it('matches with snapshot', () => {
      expect(component).toMatchSnapshot()
    })
  })

  describe('With peopleAvatar prop', () => {
    component = renderer
      .create(
        <PeopleCard>
          <PeopleCard.Image peopleAvatar />
          <PeopleCard.Cell>
            Fusce dapibus, tellus ac cursus commodo, tortor mauris condimentum
            nibh, ut fermentum massa justo sit amet risus.
          </PeopleCard.Cell>
          <PeopleCard.LastCell>•••</PeopleCard.LastCell>
        </PeopleCard>,
      )
      .toJSON()

    it('matches with snapshot', () => {
      expect(component).toMatchSnapshot()
    })
  })
})
