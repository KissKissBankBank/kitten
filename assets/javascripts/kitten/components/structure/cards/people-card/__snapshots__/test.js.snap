// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PeopleCard /> With peopleAvatar prop matches with snapshot 1`] = `
<article
  className="sc-dkPtRN qWNZV k-PeopleCard"
>
  <div
    className="k-PeopleCard__image"
  >
    <div
      className="sc-gsDKAQ knftEm k-Avatar__wrapper"
    >
      <div
        className="k-Avatar k-Avatar--small"
        style={
          Object {
            "--Avatar-height": undefined,
            "--Avatar-width": undefined,
            "--k-Avatar-background-color": "var(--color-primary-900, #0496cc)",
            "--k-Avatar-color": "var(--color-grey-000, #fff)",
          }
        }
      >
        <span
          className="k-Avatar__imageWrapper"
        >
          <img
            alt=""
            className="k-Avatar__image"
            src="/kitten-0.jpg"
          />
        </span>
      </div>
    </div>
  </div>
  <div
    className="k-PeopleCard__cell"
  >
    Fusce dapibus, tellus ac cursus commodo, tortor mauris condimentum nibh, ut fermentum massa justo sit amet risus.
  </div>
  <div
    className="k-PeopleCard__lastCell"
  >
    •••
  </div>
</article>
`;

exports[`<PeopleCard /> default matches with snapshot 1`] = `
<article
  className="sc-dkPtRN qWNZV k-PeopleCard"
>
  <div
    className="k-PeopleCard__image"
  >
    <div
      className="sc-gsDKAQ knftEm k-Avatar__wrapper"
    >
      <div
        className="k-Avatar k-Avatar--small"
        style={
          Object {
            "--Avatar-height": undefined,
            "--Avatar-width": undefined,
            "--k-Avatar-background-color": "var(--color-primary-900, #0496cc)",
            "--k-Avatar-color": "var(--color-grey-000, #fff)",
          }
        }
      >
        <span
          className="k-Avatar__imageWrapper"
        >
          <img
            alt=""
            className="k-Avatar__image"
            src="/kitten-0.jpg"
          />
        </span>
      </div>
    </div>
  </div>
  <div
    className="k-PeopleCard__cell"
  >
    Fusce dapibus, tellus ac cursus commodo, tortor mauris condimentum nibh, ut fermentum massa justo sit amet risus.
  </div>
  <div
    className="k-PeopleCard__lastCell"
  >
    •••
  </div>
</article>
`;
